whisper/timing.py

```diff
@@ -0,0 +1,198 @@
+
+    return dtw_cpu(x.double().cpu().numpy())
+
+
+def add_word_timestamps(
+    model: "Whisper",
+    tokenizer: Tokenizer,
+    mel: torch.Tensor,
+    num_frames: int,
+    segments: List[dict],
+    *,
+    medfilt_width: int = 7,
+    qk_scale: float = 1.0,
+):
+    if len(segments) == 0:
+        return
+
+    # install hooks on the cross attention layers to retrieve the attention weights
+    QKs = [None] * model.dims.n_text_layer
+    hooks = [
+        block.cross_attn.register_forward_hook(
+            lambda _, ins, outs, index=i: QKs.__setitem__(index, outs[-1])
+        )
+        for i, block in enumerate(model.decoder.blocks)
+    ]
+
+    tokens = torch.tensor(
+        [
+            *tokenizer.sot_sequence,
+            tokenizer.timestamp_begin,
+            *[t for segment in segments for t in segment["tokens"]],
+            tokenizer.timestamp_begin + mel.shape[-1] // 2,
+            tokenizer.eot,
+        ]
+    ).to(model.device)
+
+    with torch.no_grad():
+        model(mel.unsqueeze(0), tokens.unsqueeze(0))
+
+    for hook in hooks:
+        hook.remove()
+
+    weights = torch.cat(QKs[-6:])  # layers * heads * tokens * frames
+    weights = weights[:, :, :, : num_frames // 2]
+    weights = median_filter(weights, medfilt_width)
+    weights = (weights * qk_scale).softmax(dim=-1)
+    weights = weights / weights.norm(dim=-2, keepdim=True)
+    matrix = weights.mean(axis=(0, 1)).neg()
+
+    text_indices, time_indices = dtw(matrix)
+
+    jumps = np.pad(np.diff(text_indices), (1, 0), constant_values=1).astype(bool)
+    jump_times = time_indices[jumps] / TOKENS_PER_SECOND
+
+    if tokenizer.language in {"zh", "ja", "th", "lo", "my"}:
+        # These languages don't typically use spaces, so it is difficult to split words
+        # without morpheme analysis. Here, we instead split words at any
+        # position where the tokens are decoded as valid unicode points
+        split_tokens = tokenizer.split_tokens_on_unicode
+    else:
+        split_tokens = tokenizer.split_tokens_on_spaces
+
+    words, word_tokens = split_tokens(tokens[1:].tolist())
+
+    token_sources = np.repeat(np.arange(len(segments)), [len(s["tokens"]) for s in segments])
+    token_sources = [None] * len(tokenizer.sot_sequence) + list(token_sources)
+
+    time_offset = segments[0]["seek"] * HOP_LENGTH / SAMPLE_RATE
+    word_boundaries = np.pad(np.cumsum([len(t) for t in word_tokens]), (1, 0))
+    start_times = time_offset + jump_times[word_boundaries[:-1]]
+    end_times = time_offset + jump_times[word_boundaries[1:]]
+
+    for segment in segments:
+        segment["words"] = []
+
+    for i, (word, start, end) in enumerate(zip(words, start_times, end_times)):
+        if word.startswith("<|") or word.strip() in ".,!?、。":  # TODO: expand
+            continue
+
+        segment = segments[token_sources[word_boundaries[i]]]
+        segment["words"].append(dict(word=word, start=round(start, 2), end=round(end, 2)))
```

Comment:
Would it be possible to add a confidence score based on the average log proba for each words?
This can be a useful feature, available with very little additional computations.

Reply:
Great point! Added in 5fa4356

Reply:
Awesome 👍